weave/flow/eval.py

```diff
@@ -166,70 +166,78 @@ def model_post_init(self, __context: Any) -> None:
                 logger,
                 "Using 'model_output' key for compatibility with older scorers. Please update scorers to use 'output' parameter.",
             )
-        self.scorers = scorers
+
+        # I don't understand why we need a type ignore here, error:
+        # Incompatible types in assignment (expression has type "list[Op | Scorer]", variable has type "list[Callable[..., Any] | Op | Scorer] | None")
+        # This seems to be a bug in the type checker as the assignment is a valid subset of the type.
+        self.scorers = scorers  # type: ignore
 
         if isinstance(self.dataset, list):
             self.dataset = Dataset(rows=self.dataset)
 
         if self.name is None and self.dataset.name is not None:
             self.name = self.dataset.name + "-evaluation"  # type: ignore
 
+    @property
+    def _post_init_dataset(self) -> Dataset:
```

Comment:
_post_init_dataset and _post_init_scorers are a more tightly typed property. This is because the initialization code can accept lists and callables respectively, but after initialization, they are more tightly typed to the respective weave objects. Using these reduces casting below and allows us to have less logical branches